internal/sink/sink.go

```diff
@@ -0,0 +1,544 @@
+				}
+			}
+
+			batchesPerNamespace[message.Namespace] = append(batchesPerNamespace[message.Namespace], message)
+		}
+
+		for namespace, batch := range batchesPerNamespace {
+			list := []*serializer.CloudEventsKafkaPayload{}
+			for _, message := range batch {
+				list = append(list, message.Serialized)
+			}
+
+			err := s.config.Storage.BatchInsert(s.config.Context, namespace, list)
+			if err != nil {
+				// Note: a single error in batch will make the whole batch fail
+				if perr, ok := err.(*ProcessingError); ok {
+					switch perr.ProcessingControl {
+					case DEADLETTER:
+						deadletterMessages = append(deadletterMessages, batch...)
+					case DROP:
+						continue
+					default:
+						return fmt.Errorf("unknown error type: %s", err)
+					}
+				} else {
+					// Throwing and error means we will retry the whole batch again
+					return fmt.Errorf("failed to sink to storage: %s", err)
+				}
+			}
+			logger.Debug("succeeded to sink to storage", "buffer size", len(messages))
+		}
+
+		if len(deadletterMessages) > 0 {
+			err := s.deadLetter(deadletterMessages...)
+			if err != nil {
+				return fmt.Errorf("failed to deadletter messages: %s", err)
+			}
+		}
+	}
+
+	// 2. Commit Offset to Kafka
+	// Least once guarantee, if offset commit fails we will potentially process the same messages again as they are not committed yet
+	// If Redis write succeeds but Kafka commit fails we will reprocess messages without side effect as they will be dropped at duplicate check
+	// If Redis write fails we will double write them to ClickHouse and we have to relay on ClickHouse's deduplication
+	offsetStore := NewOffsetStore()
+	for _, message := range messages {
+		offsetStore.Add(message.KafkaMessage.TopicPartition)
+	}
+	offsets := offsetStore.Get()
+	if len(offsets) > 0 {
+		commitedOffsets, err := s.consumer.CommitOffsets(offsets)
+		if err != nil {
+			logger.Error("failed to commit offset to kafka", "err", err)
+			// do not return error here as we want to write Redis even if commit fails as we already saved them in ClickHouse
+		}
+		logger.Debug("succeeded to commit offset to kafka", "offsets", commitedOffsets)
+	}
+
+	// 3. Sink to Redis
+	// Least once guarantee, if Redis write fails we will accept messages with same idempotency key in future and
+	// we have to relay on ClickHouse's deduplication.
+	// Deduplicator is an optional dependency so we check if it's set
+	if s.config.Deduplicator != nil && len(dedupedMessages) > 0 {
+		dedupeItems := []dedupe.Item{}
+		for _, message := range dedupedMessages {
+			dedupeItems = append(dedupeItems, dedupe.Item{
+				Namespace: message.Namespace,
+				ID:        message.Serialized.Id,
+				Source:    message.Serialized.Source,
+			})
+		}
+
+		err := s.config.Deduplicator.Set(s.config.Context, dedupeItems...)
+		if err != nil {
+			logger.Error("failed to sink to redis", "err", err)
+			return fmt.Errorf("failed to sink to redis: %s", err)
```

Comment:
Thumb rule: if an error is returned, it should not be logged there. The parent should take care of that. If that doesn't work here, please leave a comment.